Python/Product/PythonTools/PythonTools/Environments/EnvironmentSwitcherManager.cs

```diff
@@ -0,0 +1,238 @@
+            _registryService = serviceProvider.GetComponentModel().GetService<IInterpreterRegistryService>();
+            _monitorSelection = serviceProvider.GetService(typeof(IVsMonitorSelection)) as IVsMonitorSelection;
+
+            _optionsService.DefaultInterpreterChanged += OnDefaultInterpreterChanged;
+            _registryService.InterpretersChanged += OnInterpretersChanged;
+            _monitorSelection.AdviseSelectionEvents(this, out _selectionEventsCookie);
+
+            AllFactories = Enumerable.Empty<IPythonInterpreterFactory>();
+        }
+
+        /// <summary>
+        /// Returns whether a Python project, workspace or document is currently
+        /// selected / active. Used to control the visibility of the status bar
+        /// switcher.
+        /// </summary>
+        public bool IsInPythonMode { get; private set; }
+
+        /// <summary>
+        /// Interpreter factory that is currently applicable to the current
+        /// project, workspace, or loose document. This may be null even when
+        /// <see cref="IsInPythonMode"/> is <c>true</c>, for example when no
+        /// environment is found on the machine.
+        /// </summary>
+        public IPythonInterpreterFactory CurrentFactory { get; private set; }
+
+        /// <summary>
+        /// List of factories to present to the user, based on the current
+        /// project, workspace or loose document.
+        /// </summary>
+        /// <remarks>
+        /// For a project, this is the list of environments referenced by the
+        /// project. For a loose document or workspace, this is the list of all
+        /// environments found on the machine. 
+        /// </remarks>
+        public IEnumerable<IPythonInterpreterFactory> AllFactories { get; private set; }
+
+        public PythonProjectNode Project {
+            get => _project;
+            private set {
+                if (_project != null) {
+                    _project.ActiveInterpreterChanged -= OnProjectActiveInterpreterChanged;
+                    _project.InterpreterFactoriesChanged -= OnProjectInterpreterFactoriesChanged;
+                }
+
+                _project = value;
+
+                if (_project != null) {
+                    _project.ActiveInterpreterChanged += OnProjectActiveInterpreterChanged;
+                    _project.InterpreterFactoriesChanged += OnProjectInterpreterFactoriesChanged;
+                }
+            }
+        }
+
+        public string GetSwitcherCommandKeyBinding() {
+            try {
+                return KeyBindingHelper.GetGlobalKeyBinding(GuidList.guidPythonToolsCmdSet, (int)PkgCmdIDList.cmdidViewEnvironmentStatus);
+            } catch (Exception ex) when (!ex.IsCriticalException()) {
+                // I've seen NRE thrown during VS shutdown
+                return null;
+            }
+        }
+
+        public Task SwitchToFactoryAsync(IPythonInterpreterFactory factory) {
```

Comment:
It doesn't look async, does it have to be?

Reply:
I did it in anticipation of OpenFolder, where the methods we'll need to use to write the settings are all async

Reply:
The OpenFolder scenario is in now.